networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py

```diff
@@ -212,64 +213,77 @@ def _add_floating_ip(self, ri, ex_gw_port, floating_ip, fixed_ip):
+        subnet = self._get_matching_subnet(subnets, floating_ip)
+        if subnet:
+            secondary_ip = netaddr.IPAddress(subnet.value +
+                                             (subnet.hostmask.value - 1))
+            self._asr_do_add_secondary_ip(secondary_ip,
+                                          ex_gw_port, subnet.netmask)
+
     def _remove_floating_ip(self, ri, ext_gw_port, floating_ip, fixed_ip):
         vrf_name = self._get_vrf_name(ri)
         self._asr_do_remove_floating_ip(ri, floating_ip,
                                         fixed_ip,
                                         vrf_name,
                                         ext_gw_port)
-
-    def _asr_do_add_secondary_ip(self, ri, floating_ip, port):
-        # get the list of extra subnets from the GW port and check
-        # if this IP address matches one of them
-        fip = netaddr.IPAddress(floating_ip)
-        secondary_ip = ''
+        # A secondary IP address may need to be removed from the external
+        # interface. Check the known subnets to see which one contains
+        # the floating IP, then search for any other floating IPs on that
+        # subnet. If there aren't any, then the secondary IP can safely
+        # be removed.
         subnets = ri.router['gw_port'].get('extra_subnets', [])
-        for subnet in subnets:
-            net = netaddr.IPNetwork(subnet['cidr'])
-            if (fip.value & net.netmask.value) == net.value:
-                secondary_ip = netaddr.IPAddress(net.value +
-                    (net.hostmask.value - 1))
-                break
-        if secondary_ip == '':
+        subnet = self._get_matching_subnet(subnets, floating_ip)
+        if not subnet:
             return
-        sub_interface = self._get_interface_name_from_hosting_port(port)
-        conf_str = (snippets.ADD_SECONDARY_IP % (
-                       sub_interface, secondary_ip, str(net.netmask)))
-        self._edit_running_config(conf_str, 'ADD_SECONDARY_IP')
-
-    def _asr_do_remove_secondary_ip(self, ri, floating_ip, port):
-        fip = netaddr.IPAddress(floating_ip)
-        secondary_ip = ''
-        # Find the subnet this belnogs to, if present
-        subnets = ri.router['gw_port'].get('extra_subnets', [])
-        for subnet in subnets:
-            net = netaddr.IPNetwork(subnet['cidr'])
-            if (fip.value & net.netmask.value) == net.value:
-                secondary_ip = netaddr.IPAddress(net.value +
-                    (net.hostmask.value - 1))
-                break
-        if secondary_ip == '':
-            return
-
-        # See if there are any other FIPs on this subnet
+        secondary_ip = netaddr.IPAddress(subnet.value +
+                                         (subnet.hostmask.value - 1))
+        # We only remove the secondary IP if there aren't any more FIPs
+        # for this subnet
+        # FIXME(tbachman): should check across all routers
```

Comment:
Pls file a support issue in github for this as we need to solve this before release.